### DISCUSSION TITLE
For the project which mixing Rust code and Python code, how the namespace looks like?

### DISCUSSION THREAD

**Discussion by evan0greenup** [PyO3/maturin#1426] [Q&A] [answered]
According to https://github.com/PyO3/maturin#readme , 

The directory layout of a PyO3 project is like this:

```
my-project
├── Cargo.toml
├── my_project
│   ├── __init__.py
│   └── bar.py
├── pyproject.toml
├── README.md
└── src
    └── lib.rs
```

or this 

```
my-project
├── Cargo.toml
├── python
│   └── my_project
│       ├── __init__.py
│       └── bar.py
├── pyproject.toml
├── README.md
└── src
    └── lib.rs
```

No matter which one, after packaging and import the root module of this package. How its name space looks like (what members does it contain?) 

Is it the Union of both python code and rust code?

Is there any way to manage the namespace?

**Comment by davidhewitt** [comment, marked as answer]
Transferred to `maturin` as this is specific to project layout when using maturin.

When using mixed Python / Rust in this way the `my_project/__init__.py` will control exactly the namespace you want. 

`maturin` will package your Rust module beside this one, e.g. if your Rust library is `my_project` then you can import from it in the`__init__.py` as if you had `my_project/my_project.py`.

(Actually because it is compiled code, the file will be `my_project.so`or similar depending on OS but it functions the same.)
